src/backend/puppet/observed.rs

```diff
@@ -0,0 +1,585 @@
+//! Structs representing output of puppet backend
+use enumset::EnumSet;
+use std::ops::Index;
+use std::ops::IndexMut;
+use std::rc::Rc;
+use std::string::ToString;
+use crate::theme::ColorPair;
+use crate::theme::Effect;
+use unicode_segmentation::UnicodeSegmentation;
+use unicode_width::UnicodeWidthStr;
+use crate::Vec2;
+
+/// Style of observed cell
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ObservedStyle {
+    /// Colors: front and back
+    pub colors: ColorPair,
+    /// Effects enabled on observed cell
+    pub effects: EnumSet<Effect>,
+}
+
+/// Contents of observed cell
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum GraphemePart {
+    /// Represents begin of wide character
+    Begin(String),
+    /// Represents a cell that is filled with continuation of some character that begun in cell with lower x-index.
+    Continuation,
+}
+
+impl GraphemePart {
+    /// Returns true iff GraphemePart is Continuation
+    pub fn is_continuation(&self) -> bool {
+        match self {
+            &GraphemePart::Continuation => true,
+            _ => false,
+        }
+    }
+
+    /// Returns Some(String) if GraphemePart is Begin(String), else None.
+    pub fn as_option(&self) -> Option<&String> {
+        match self {
+            &GraphemePart::Begin(ref string) => Some(string),
+            &GraphemePart::Continuation => None,
+        }
+    }
+
+    /// Returns String if GraphemePart is Begin(String), panics otherwise.
+    pub fn unwrap(&self) -> String {
+        match self {
+            &GraphemePart::Begin(ref s) => s.clone(),
+            _ => panic!("unwrapping GraphemePart::Continuation"),
+        }
+    }
+}
+
+/// Represents a single cell of terminal.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ObservedCell {
+    /// Absolute position
+    pub pos: Vec2,
+    /// Style
+    pub style: Rc<ObservedStyle>,
+    /// Part of grapheme - either it's beginning or continuation when character is multi-cell long.
+    pub letter: GraphemePart,
+}
+
+impl ObservedCell {
+    /// Constructor
+    pub fn new(
+        pos: Vec2,
+        style: Rc<ObservedStyle>,
+        letter: Option<String>,
+    ) -> Self {
+        let letter: GraphemePart = match letter {
+            Some(s) => GraphemePart::Begin(s),
+            None => GraphemePart::Continuation,
+        };
+
+        ObservedCell { pos, style, letter }
+    }
+}
+
+/// Puppet backend output
+///
+/// Represents single frame.
+#[derive(Debug, Clone)]
+pub struct ObservedScreen {
+    /// Size
+    size: Vec2,
+    /// Contents. Each cell can be set or empty.
+    contents: Vec<Option<ObservedCell>>,
+}
+
+impl ObservedScreen {
+    /// Creates empty ObservedScreen
+    pub fn new(size: Vec2) -> Self {
+        let contents: Vec<Option<ObservedCell>> = vec![None; size.x * size.y];
+
+        ObservedScreen { size, contents }
+    }
+
+    fn flatten_index(&self, index: &Vec2) -> usize {
```

Comment:
`Vec2` is copy, in general it's easier to pass-by-value.

Reply:
fixed it everywhere, thanks!